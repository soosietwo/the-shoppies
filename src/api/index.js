const getMovies = async (searchTerm, page = 1) => {
  let url = `https://www.omdbapi.com/?apikey=${
    process.env.REACT_APP_API_KEY
  }&type=movie&s=${searchTerm.trim()}&page=${page}`;

  return (await fetch(url)).json();
};

export default {
  getMovies,
};
