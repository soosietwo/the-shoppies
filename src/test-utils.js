// test-utils.js
import React from "react";
import { render } from "@testing-library/react";
import { AppProvider, Frame } from "@shopify/polaris";
import en from "@shopify/polaris/locales/en.json";
import Store from "./store";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const AllTheProviders = ({ children }) => {
  return (
    <AppProvider i18n={en}>
      <Store>
        <Frame>{children}</Frame>
      </Store>
    </AppProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
