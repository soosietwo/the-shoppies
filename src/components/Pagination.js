import React, { useContext } from "react";
import { Pagination } from "@shopify/polaris";

import { Context } from "../store";
import api from "../api";
import {
  FETCH_NEXT,
  FETCH_PREVIOUS,
  FETCH_MOVIES_SUCCESS,
  FETCH_MOVIES_FAILURE,
} from "../store/constants";

const PaginationComponent = () => {
  const [state, dispatch] = useContext(Context);
  const totalPages = Math.ceil(state.totalResults / 10);

  const getPage = (page) => {
    dispatch({ type: page > state.currentPage ? FETCH_NEXT : FETCH_PREVIOUS });
    api
      .getMovies(state.searchTerm, null, page)
      .then(({ Error, Search, totalResults }) => {
        if (Error) {
          dispatch({ type: FETCH_MOVIES_FAILURE, payload: Error });
        } else {
          dispatch({
            type: FETCH_MOVIES_SUCCESS,
            payload: { Search, totalResults },
          });
        }
      })
      .catch((error) => {
        dispatch({ type: FETCH_MOVIES_FAILURE, payload: error });
      });
  };

  return (
    <Pagination
      label={`Page ${state.currentPage} of ${totalPages}`}
      hasPrevious={state.currentPage > 1}
      onPrevious={() => {
        getPage(state.currentPage - 1);
      }}
      hasNext={state.currentPage < totalPages}
      onNext={() => {
        getPage(state.currentPage + 1);
      }}
    />
  );
};

export default PaginationComponent;
