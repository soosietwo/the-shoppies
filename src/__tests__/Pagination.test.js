import React from "react";
import { render, waitFor } from "../test-utils";
import Pagination from "../components/Pagination";
import { fireEvent } from "@testing-library/react/dist/pure";
import api from "../api";
import Store from "../store";

describe("Pagination", () => {
  it("renders and matches snapshot", () => {
    const mockInitialState = {
      currentPage: 1,
      totalResults: 10,
    };

    const { container } = render(
      <Store initialState={mockInitialState}>
        <Pagination />
      </Store>
    );

    expect(container.querySelector(".Polaris-Pagination")).toMatchSnapshot();
  });

  it("calls getMovies when you click next or previous ", async () => {
    const mockInitialState = {
      currentPage: 1,
      totalResults: 50,
    };

    const { container } = render(
      <Store initialState={mockInitialState}>
        <Pagination />
      </Store>
    );

    const getMoviesSpy = jest.spyOn(api, "getMovies");

    fireEvent.click(container.querySelector('[aria-label="Next"]'));

    await waitFor(() => {
      expect(getMoviesSpy).toHaveBeenCalledWith(undefined, 2);
    });

    fireEvent.click(container.querySelector('[aria-label="Previous"]'));

    await waitFor(() => {
      expect(getMoviesSpy).toHaveBeenCalledWith(undefined, 1);
    });
  });

  it("disables the previous button on the first page", async () => {
    const mockInitialState = {
      currentPage: 1,
      totalResults: 20,
    };

    const { container } = render(
      <Store initialState={mockInitialState}>
        <Pagination />
      </Store>
    );

    expect(container.querySelector('[aria-label="Previous"]')).toBeDisabled();
  });

  it("disables the next button on the last page", async () => {
    const mockInitialState = {
      currentPage: 5,
      totalResults: 50,
    };

    const { container } = render(
      <Store initialState={mockInitialState}>
        <Pagination />
      </Store>
    );

    expect(container.querySelector('[aria-label="Next"]')).toBeDisabled();
  });
});
