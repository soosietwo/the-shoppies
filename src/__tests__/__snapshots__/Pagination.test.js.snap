// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders and matches snapshot 1`] = `
<nav
  aria-label="Pagination"
  class="Polaris-Pagination"
>
  <button
    aria-label="Previous"
    class="Polaris-Pagination__Button"
    disabled=""
    type="button"
  >
    <span
      class="Polaris-Icon"
    >
      <svg
        aria-hidden="true"
        class="Polaris-Icon__Svg"
        focusable="false"
        viewBox="0 0 20 20"
      >
        <path
          d="M17 9H5.414l3.293-3.293a.999.999 0 1 0-1.414-1.414l-5 5a.999.999 0 0 0 0 1.414l5 5a.997.997 0 0 0 1.414 0 .999.999 0 0 0 0-1.414L5.414 11H17a1 1 0 1 0 0-2z"
        />
      </svg>
    </span>
  </button>
  <div
    aria-live="polite"
    class="Polaris-Pagination__Label"
  >
    <span
      class="Polaris-TextStyle--variationSubdued"
    >
      Page 1 of 1
    </span>
  </div>
  <button
    aria-label="Next"
    class="Polaris-Pagination__Button"
    disabled=""
    type="button"
  >
    <span
      class="Polaris-Icon"
    >
      <svg
        aria-hidden="true"
        class="Polaris-Icon__Svg"
        focusable="false"
        viewBox="0 0 20 20"
      >
        <path
          d="M17.707 9.293l-5-5a.999.999 0 1 0-1.414 1.414L14.586 9H3a1 1 0 1 0 0 2h11.586l-3.293 3.293a.999.999 0 1 0 1.414 1.414l5-5a.999.999 0 0 0 0-1.414z"
        />
      </svg>
    </span>
  </button>
</nav>
`;
